File: win/src/KelvinShift/Services/SettingsService.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.IO;
using System.Runtime.CompilerServices;
using System.Text.Json;
using System.Windows.Threading;

namespace KelvinShift.Services;

// Mirrors macos/Sources/KelvinShift/Settings.swift. Persists to
// %APPDATA%\KelvinShift\settings.json with a 250ms debounced write so
// dragging sliders doesn't hammer disk. Raises SettingsChanged on every
// property change so the schedule engine can re-tick.
public sealed class SettingsService : INotifyPropertyChanged
{
    public event PropertyChangedEventHandler? PropertyChanged;
    public event EventHandler? SettingsChanged;

    private readonly string _path;
    private bool _loading;
    private DispatcherTimer? _saveDebounce;

    public SettingsService()
    {
        var dir = Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData),
            "KelvinShift");
        Directory.CreateDirectory(dir);
        _path = Path.Combine(dir, "settings.json");
    }

    // ── Color Temperature ──────────────────────────────────
    private int _dayKelvin = 5000;
    public int DayKelvin { get => _dayKelvin; set => Set(ref _dayKelvin, Math.Clamp(value, 2000, 6500)); }

    private int _nightKelvin = 2700;
    public int NightKelvin { get => _nightKelvin; set => Set(ref _nightKelvin, Math.Clamp(value, 1800, 5500)); }

    // ── Brightness ────────────────────────────────────────
    private double _dayBrightness = 1.0;
    public double DayBrightness { get => _dayBrightness; set => Set(ref _dayBrightness, Math.Clamp(value, 0.1, 1.0)); }

    private double _nightBrightness = 0.8;
    public double NightBrightness { get => _nightBrightness; set => Set(ref _nightBrightness, Math.Clamp(value, 0.1, 1.0)); }

    // ── Schedule ───────────────────────────────────────────
    private string _scheduleMode = "custom";
    public string ScheduleMode { get => _scheduleMode; set => Set(ref _scheduleMode, value); }

    private int _customDayHour = 7;
    public int CustomDayHour { get => _customDayHour; set => Set(ref _customDayHour, Math.Clamp(value, 0, 23)); }

    private int _customDayMinute = 0;
    public int CustomDayMinute { get => _customDayMinute; set => Set(ref _customDayMinute, Math.Clamp(value, 0, 59)); }

    private int _customNightHour = 20;
    public int CustomNightHour { get => _customNightHour; set => Set(ref _customNightHour, Math.Clamp(value, 0, 23)); }

    private int _customNightMinute = 0;
    public int CustomNightMinute { get => _customNightMinute; set => Set(ref _customNightMinute, Math.Clamp(value, 0, 59)); }

    // ── Location ──────────────────────────────────────────
    private double _latitude;
    public double Latitude { get => _latitude; set => Set(ref _latitude, value); }

    private double _longitude;
    public double Longitude { get => _longitude; set => Set(ref _longitude, value); }

    private string _locationName = "";
    public string LocationName { get => _locationName; set => Set(ref _locationName, value); }

    // ── Transition ─────────────────────────────────────────
    private int _transitionMinutes = 20;
    public int TransitionMinutes { get => _transitionMinutes; set => Set(ref _transitionMinutes, Math.Max(1, value)); }

    // ── Bedtime ───────────────────────────────────────────
    private bool _bedtimeEnabled;
    public bool BedtimeEnabled { get => _bedtimeEnabled; set => Set(ref _bedtimeEnabled, value); }

    private int _bedtimeKelvin = 1900;
    public int BedtimeKelvin { get => _bedtimeKelvin; set => Set(ref _bedtimeKelvin, Math.Clamp(value, 1000, 4000)); }

    private double _bedtimeBrightness = 0.4;
    public double BedtimeBrightness { get => _bedtimeBrightness; set => Set(ref _bedtimeBrightness, Math.Clamp(value, 0.1, 1.0)); }

    private int _bedtimeHour = 23;
    public int BedtimeHour { get => _bedtimeHour; set => Set(ref _bedtimeHour, Math.Clamp(value, 0, 23)); }

    private int _bedtimeMinute;
    public int BedtimeMinute { get => _bedtimeMinute; set => Set(ref _bedtimeMinute, Math.Clamp(value, 0, 59)); }

    private int _bedtimeRampMinutes = 60;
    public int BedtimeRampMinutes { get => _bedtimeRampMinutes; set => Set(ref _bedtimeRampMinutes, Math.Max(1, value)); }

    // ── Master toggle ─────────────────────────────────────
    private bool _enabled = true;
    public bool Enabled { get => _enabled; set => Set(ref _enabled, value); }

    // ── System color pipeline (MHC) ───────────────────────
    // Opt-out toggle for the GPU hardware color transform path. When on
    // (default) and the GPU supports it, color goes through Windows' MHC
    // pipeline — no Action Center flash, no screenshot tint. Off falls
    // back to the gamma ramp path (some flash on shell flyouts possible).
    private bool _useSystemColorPipeline = true;
    public bool UseSystemColorPipeline { get => _useSystemColorPipeline; set => Set(ref _useSystemColorPipeline, value); }

    // ── Launch at Login ───────────────────────────────────
    private bool _launchAtLogin;
    public bool LaunchAtLogin
    {
        get => _launchAtLogin;
        set
        {
            if (Set(ref _launchAtLogin, value))
                LaunchAtLoginService.Apply(value);
        }
    }

    // ── Persistence ───────────────────────────────────────

    public void Load()
    {
        if (!File.Exists(_path)) return;
        try
        {
            _loading = true;
            var dto = JsonSerializer.Deserialize<Dto>(File.ReadAllText(_path));
            if (dto is null) return;
            DayKelvin         = dto.DayKelvin;
            NightKelvin       = dto.NightKelvin;
            DayBrightness     = dto.DayBrightness;
            NightBrightness   = dto.NightBrightness;
            ScheduleMode      = dto.ScheduleMode ?? "custom";
            CustomDayHour     = dto.CustomDayHour;
            CustomDayMinute   = dto.CustomDayMinute;
            CustomNightHour   = dto.CustomNightHour;
            CustomNightMinute = dto.CustomNightMinute;
            Latitude          = dto.Latitude;
            Longitude         = dto.Longitude;
            LocationName      = dto.LocationName ?? "";
            TransitionMinutes = dto.TransitionMinutes == 0 ? 20 : dto.TransitionMinutes;
            BedtimeEnabled    = dto.BedtimeEnabled;
            BedtimeKelvin     = dto.BedtimeKelvin == 0 ? 1900 : dto.BedtimeKelvin;
            BedtimeBrightness = dto.BedtimeBrightness == 0 ? 0.4 : dto.BedtimeBrightness;
            BedtimeHour       = dto.BedtimeHour;
            BedtimeMinute     = dto.BedtimeMinute;
            BedtimeRampMinutes = dto.BedtimeRampMinutes == 0 ? 60 : dto.BedtimeRampMinutes;
            UseSystemColorPipeline = dto.UseSystemColorPipeline;
            Enabled           = dto.Enabled;
            LaunchAtLogin     = LaunchAtLoginService.IsEnabled();
        }
        catch { /* leave defaults */ }
        finally { _loading = false; }
    }

    private void Save()
    {
        var dto = new Dto
        {
            DayKelvin = DayKelvin, NightKelvin = NightKelvin,
            DayBrightness = DayBrightness, NightBrightness = NightBrightness,
            ScheduleMode = ScheduleMode,
            CustomDayHour = CustomDayHour, CustomDayMinute = CustomDayMinute,
            CustomNightHour = CustomNightHour, CustomNightMinute = CustomNightMinute,
            Latitude = Latitude, Longitude = Longitude, LocationName = LocationName,
            TransitionMinutes = TransitionMinutes,
            BedtimeEnabled = BedtimeEnabled,
            BedtimeKelvin = BedtimeKelvin, BedtimeBrightness = BedtimeBrightness,
            BedtimeHour = BedtimeHour, BedtimeMinute = BedtimeMinute,
            BedtimeRampMinutes = BedtimeRampMinutes,
            UseSystemColorPipeline = UseSystemColorPipeline,
            Enabled = Enabled,
        };
        File.WriteAllText(_path, JsonSerializer.Serialize(dto, new JsonSerializerOptions { WriteIndented = true }));
    }

    private void ScheduleSave()
    {
        _saveDebounce ??= new DispatcherTimer { Interval = TimeSpan.FromMilliseconds(250) };
        _saveDebounce.Stop();
        _saveDebounce.Tick -= OnSaveTick;
        _saveDebounce.Tick += OnSaveTick;
        _saveDebounce.Start();
    }
    private void OnSaveTick(object? s, EventArgs e) { _saveDebounce!.Stop(); Save(); }

    private bool Set<T>(ref T field, T value, [CallerMemberName] string? name = null)
    {
        if (EqualityComparer<T>.Default.Equals(field, value)) return false;
        field = value;
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(name));
        if (!_loading)
        {
            ScheduleSave();
            SettingsChanged?.Invoke(this, EventArgs.Empty);
        }
        return true;
    }

    public string DayTimeLabel     => FormatTime(CustomDayHour, CustomDayMinute);
    public string NightTimeLabel   => FormatTime(CustomNightHour, CustomNightMinute);
    public string BedtimeTimeLabel => FormatTime(BedtimeHour, BedtimeMinute);

    private static string FormatTime(int h, int m)
    {
        var h12 = h == 0 ? 12 : (h > 12 ? h - 12 : h);
        var sfx = h >= 12 ? "PM" : "AM";
        return $"{h12}:{m:D2} {sfx}";
    }

    private sealed class Dto
    {
        public int DayKelvin { get; set; }
        public int NightKelvin { get; set; }
        public double DayBrightness { get; set; }
        public double NightBrightness { get; set; }
        public string? ScheduleMode { get; set; }
        public int CustomDayHour { get; set; }
        public int CustomDayMinute { get; set; }
        public int CustomNightHour { get; set; }
        public int CustomNightMinute { get; set; }
        public double Latitude { get; set; }
        public double Longitude { get; set; }
        public string? LocationName { get; set; }
        public int TransitionMinutes { get; set; }
        public bool BedtimeEnabled { get; set; }
        public int BedtimeKelvin { get; set; }
        public double BedtimeBrightness { get; set; }
        public int BedtimeHour { get; set; }
        public int BedtimeMinute { get; set; }
        public int BedtimeRampMinutes { get; set; }
        public bool UseSystemColorPipeline { get; set; } = true;
        public bool Enabled { get; set; }
    }
}

File: win32/src/SettingsService.cpp
#include "SettingsService.h"
#include "Json.h"
#include "LaunchAtLoginService.h"
#include <cstdio>

SettingsService::SettingsService()
{
    std::wstring dir = KnownFolder(FOLDERID_RoamingAppData) + L"\\KelvinShift";
    CreateDirectoryW(dir.c_str(), nullptr);
    path_ = dir + L"\\settings.json";

    // One-shot debounce: each change restarts it; the tick saves once.
    saveDebounce_.SetInterval(250);
    saveDebounce_.SetCallback([this] { saveDebounce_.Stop(); Save(); });
}

// ── Change plumbing ───────────────────────────────────────────────────────

template <class T>
bool SettingsService::Set(T& field, const T& value, const char* name)
{
    if (field == value) return false;
    field = value;
    PropertyChanged.Raise(std::string(name));
    if (!loading_)
    {
        ScheduleSave();
        SettingsChanged.Raise();
    }
    return true;
}

// Explicit instantiations for the types actually used.
template bool SettingsService::Set<int>(int&, const int&, const char*);
template bool SettingsService::Set<double>(double&, const double&, const char*);
template bool SettingsService::Set<bool>(bool&, const bool&, const char*);
template bool SettingsService::Set<std::string>(std::string&, const std::string&, const char*);

// ── Setters (clamping mirrors SettingsService.cs) ─────────────────────────

void SettingsService::DayKelvin(int v)        { Set(dayKelvin_, std::clamp(v, 1000, 6500), "DayKelvin"); }
void SettingsService::NightKelvin(int v)      { Set(nightKelvin_, std::clamp(v, 1000, 6500), "NightKelvin"); }
void SettingsService::DayBrightness(double v) { Set(dayBrightness_, std::clamp(v, 0.1, 1.0), "DayBrightness"); }
void SettingsService::NightBrightness(double v){ Set(nightBrightness_, std::clamp(v, 0.1, 1.0), "NightBrightness"); }
void SettingsService::ScheduleMode(const std::string& v) { Set(scheduleMode_, v, "ScheduleMode"); }
void SettingsService::CustomDayHour(int v)    { Set(customDayHour_, std::clamp(v, 0, 23), "CustomDayHour"); }
void SettingsService::CustomDayMinute(int v)  { Set(customDayMinute_, std::clamp(v, 0, 59), "CustomDayMinute"); }
void SettingsService::CustomNightHour(int v)  { Set(customNightHour_, std::clamp(v, 0, 23), "CustomNightHour"); }
void SettingsService::CustomNightMinute(int v){ Set(customNightMinute_, std::clamp(v, 0, 59), "CustomNightMinute"); }
void SettingsService::Latitude(double v)      { Set(latitude_, v, "Latitude"); }
void SettingsService::Longitude(double v)     { Set(longitude_, v, "Longitude"); }
void SettingsService::LocationName(const std::string& v) { Set(locationName_, v, "LocationName"); }
void SettingsService::TransitionMinutes(int v){ Set(transitionMinutes_, std::max(1, v), "TransitionMinutes"); }
void SettingsService::BedtimeEnabled(bool v)  { Set(bedtimeEnabled_, v, "BedtimeEnabled"); }
void SettingsService::BedtimeKelvin(int v)    { Set(bedtimeKelvin_, std::clamp(v, 1000, 6500), "BedtimeKelvin"); }
void SettingsService::BedtimeBrightness(double v){ Set(bedtimeBrightness_, std::clamp(v, 0.1, 1.0), "BedtimeBrightness"); }
void SettingsService::BedtimeHour(int v)      { Set(bedtimeHour_, std::clamp(v, 0, 23), "BedtimeHour"); }
void SettingsService::BedtimeMinute(int v)    { Set(bedtimeMinute_, std::clamp(v, 0, 59), "BedtimeMinute"); }
void SettingsService::BedtimeRampMinutes(int v){ Set(bedtimeRampMinutes_, std::max(1, v), "BedtimeRampMinutes"); }
void SettingsService::Enabled(bool v)         { Set(enabled_, v, "Enabled"); }
void SettingsService::UseSystemColorPipeline(bool v) { Set(useSystemColorPipeline_, v, "UseSystemColorPipeline"); }

void SettingsService::LaunchAtLogin(bool v)
{
    if (Set(launchAtLogin_, v, "LaunchAtLogin"))
        LaunchAtLoginService::Apply(v);
}

// ── Persistence ───────────────────────────────────────────────────────────

void SettingsService::Load()
{
    HANDLE f = CreateFileW(path_.c_str(), GENERIC_READ, FILE_SHARE_READ, nullptr,
                           OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL, nullptr);
    if (f == INVALID_HANDLE_VALUE) return; // first run — keep defaults

    std::string text;
    char buf[4096];
    DWORD got;
    while (ReadFile(f, buf, sizeof(buf), &got, nullptr) && got > 0)
        text.append(buf, got);
    CloseHandle(f);

    JsonReader r;
    if (!r.Parse(text)) return;

    loading_ = true;
    DayKelvin(r.GetInt("DayKelvin", dayKelvin_));
    NightKelvin(r.GetInt("NightKelvin", nightKelvin_));
    DayBrightness(r.GetDouble("DayBrightness", dayBrightness_));
    NightBrightness(r.GetDouble("NightBrightness", nightBrightness_));
    ScheduleMode(r.GetString("ScheduleMode", "custom"));
    CustomDayHour(r.GetInt("CustomDayHour", customDayHour_));
    CustomDayMinute(r.GetInt("CustomDayMinute", customDayMinute_));
    CustomNightHour(r.GetInt("CustomNightHour", customNightHour_));
    CustomNightMinute(r.GetInt("CustomNightMinute", customNightMinute_));
    Latitude(r.GetDouble("Latitude", 0.0));
    Longitude(r.GetDouble("Longitude", 0.0));
    LocationName(r.GetString("LocationName", ""));
    {
        // 0 (or missing) is treated as "unset" -> default, matching the C# DTO.
        int tm = r.GetInt("TransitionMinutes", 0);
        TransitionMinutes(tm == 0 ? 20 : tm);
    }
    BedtimeEnabled(r.GetBool("BedtimeEnabled", false));
    {
        int bk = r.GetInt("BedtimeKelvin", 0);
        BedtimeKelvin(bk == 0 ? 1900 : bk);
    }
    {
        double bb = r.GetDouble("BedtimeBrightness", 0.0);
        BedtimeBrightness(bb == 0.0 ? 0.4 : bb);
    }
    BedtimeHour(r.GetInt("BedtimeHour", bedtimeHour_));
    BedtimeMinute(r.GetInt("BedtimeMinute", 0));
    {
        int br = r.GetInt("BedtimeRampMinutes", 0);
        BedtimeRampMinutes(br == 0 ? 60 : br);
    }
    UseSystemColorPipeline(r.GetBool("UseSystemColorPipeline", true));
    Enabled(r.GetBool("Enabled", true));
    loading_ = false;

    // Launch-at-login is sourced from the registry, not the file.
    launchAtLogin_ = LaunchAtLoginService::IsEnabled();
}

void SettingsService::Save()
{
    JsonWriter w;
    w.Int("DayKelvin", dayKelvin_);
    w.Int("NightKelvin", nightKelvin_);
    w.Double("DayBrightness", dayBrightness_);
    w.Double("NightBrightness", nightBrightness_);
    w.Str("ScheduleMode", scheduleMode_);
    w.Int("CustomDayHour", customDayHour_);
    w.Int("CustomDayMinute", customDayMinute_);
    w.Int("CustomNightHour", customNightHour_);
    w.Int("CustomNightMinute", customNightMinute_);
    w.Double("Latitude", latitude_);
    w.Double("Longitude", longitude_);
    w.Str("LocationName", locationName_);
    w.Int("TransitionMinutes", transitionMinutes_);
    w.Bool("BedtimeEnabled", bedtimeEnabled_);
    w.Int("BedtimeKelvin", bedtimeKelvin_);
    w.Double("BedtimeBrightness", bedtimeBrightness_);
    w.Int("BedtimeHour", bedtimeHour_);
    w.Int("BedtimeMinute", bedtimeMinute_);
    w.Int("BedtimeRampMinutes", bedtimeRampMinutes_);
    w.Bool("UseSystemColorPipeline", useSystemColorPipeline_);
    w.Bool("Enabled", enabled_);
    std::string json = w.Done();

    // Write to a temp file then rename — never leave a half-written settings
    // file if the process dies mid-save.
    std::wstring tmp = path_ + L".tmp";
    HANDLE f = CreateFileW(tmp.c_str(), GENERIC_WRITE, 0, nullptr,
                           CREATE_ALWAYS, FILE_ATTRIBUTE_NORMAL, nullptr);
    if (f == INVALID_HANDLE_VALUE) return;
    DWORD written;
    WriteFile(f, json.data(), (DWORD)json.size(), &written, nullptr);
    CloseHandle(f);
    MoveFileExW(tmp.c_str(), path_.c_str(), MOVEFILE_REPLACE_EXISTING);
}

void SettingsService::ScheduleSave()
{
    saveDebounce_.Start(); // Start() restarts from zero — the debounce
}

// ── Labels ────────────────────────────────────────────────────────────────

static std::string FormatTime(int h, int m)
{
    int h12 = (h == 0) ? 12 : (h > 12 ? h - 12 : h);
    const char* sfx = (h >= 12) ? "PM" : "AM";
    char buf[16];
    snprintf(buf, sizeof(buf), "%d:%02d %s", h12, m, sfx);
    return buf;
}

std::string SettingsService::DayTimeLabel() const   { return FormatTime(customDayHour_, customDayMinute_); }
std::string SettingsService::NightTimeLabel() const { return FormatTime(customNightHour_, customNightMinute_); }
std::string SettingsService::BedtimeTimeLabel() const { return FormatTime(bedtimeHour_, bedtimeMinute_); }
